### ISSUE TITLE
Service Workers in web browsers with stricter restrictions

### ISSUE THREAD

**Issue by jtpio** [jupyterlite/jupyterlite#679]
## Description

At the moment using service workers in a Firefox private window is a known issue.

However there seems to be other cases where the service worker would fail to activate.

For example with Firefox with "Enhanced Tracking Protection":

[![image](https://user-images.githubusercontent.com/591645/174609127-483bd7fe-c2f8-4571-a772-c2da79d145be.png)](https://user-images.githubusercontent.com/591645/174609127-483bd7fe-c2f8-4571-a772-c2da79d145be.png)

This also seems to be the case with other flavors of Chromium like `ungoogled-chromium`: https://github.com/ungoogled-software/ungoogled-chromium

## Reproduce

Open https://jupyterlite.rtfd.io/en/latest/try/lab in one of the browsers mentioned above.

## Expected behavior

Opening the issue to see if there is something to be done to handle these cases (or improve error handling).

## Context

- JupyterLite version: 0.1.0b9
- Browser and version: Firefox Developer Edition `102.0b9`

**Comment by bollwyvl** [comment]
Yep. There are a large number of issues with the serviceworker approach, some of which have been known as long as #137 (and we aren't even getting the offline benefits of that yet!) As we've learned, there's a huge gulf between a naive assumption of "most deployments on most browsers" and the specific deployment someone has in mind.

Much as with chrome-only features, we could put the entire service worker feature behind a separate installable package, just leaving hooks here, or at least a put it behind a configuration flag: this way, a site deployer would at least need to do some reading before declaring victory at "works for me on `localhost`".

For the (pyodide) emscripten kernel contents in specific: at some point in the future, there will be an emscripten-native way for us to handle the async nature of our contents manager, without a service worker, so that could move that back into core.

**Comment by bollwyvl** [comment]
Semi-unrelated: a recent development seems to start spinning these messages every couple seconds trying to poll for WebRTC signaling servers:

![Screenshot from 2022-06-22 08-18-41](https://user-images.githubusercontent.com/45380/175038817-d3ae8714-4717-4186-9248-67b0eaccb7d1.png)

**Comment by bollwyvl** [comment]
see also https://github.com/jupyterlite/jupyterlite/issues/712

**Comment by jtpio** [comment]
With https://github.com/jupyterlite/jupyterlite/issues/1498 the kernels could now log a warning in the log console if the service worker is not available and not `crossOriginIsolated`.

This would also likely fix https://github.com/jupyterlite/jupyterlite/issues/1432

**Comment by jtpio** [comment]
Ah, it looks like Service Workers would start working in private windows with the next Firefox release! (testing with Firefox Dev Edition - `140.0b4`)

https://github.com/user-attachments/assets/def6d3d2-ca20-4ef9-9aee-ae1e97732a42

**Comment by bollwyvl** [comment]
Yeah: and according to the proposed release schedule,
sw-in-private-browsing should be in the next ESR in September.
